src/librustc_mir/interpret/intrinsics.rs

```diff
@@ -220,7 +221,29 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
             sym::discriminant_value => {
                 let place = self.deref_operand(args[0])?;
                 let discr_val = self.read_discriminant(place.into())?.0;
-                self.write_scalar(Scalar::from_u64(u64::try_from(discr_val).unwrap()), dest)?;
+                let scalar = match dest.layout.ty.kind {
+                    ty::Int(IntTy::Isize) => {
+                        Scalar::from_machine_isize(i64::try_from(discr_val as i128).unwrap(), self)
+                    }
+                    ty::Uint(UintTy::Usize) => {
+                        Scalar::from_machine_usize(u64::try_from(discr_val).unwrap(), self)
+                    }
+                    ty::Int(int) => Scalar::from_int(
+                        discr_val as i128,
+                        Size::from_bits(int.bit_width().unwrap()),
```

Comment:
You can use `dest.layout.size`. And that means it will support `isize` too. (same for below and `usize`)